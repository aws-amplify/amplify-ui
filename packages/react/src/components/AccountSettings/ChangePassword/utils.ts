import { FormValues, ValidationError } from '../types';

export const getIsDisabled = (
  formValues: FormValues,
  validationError: ValidationError
): boolean => {
  const { newPassword, confirmPassword } = formValues;

  if (!newPassword || !confirmPassword) {
    // if passwords aren't entered yet, disable submit
    return true;
  } else if (
    // if there are some password validation error, disable submit
    validationError.newPassword?.length > 0 ||
    validationError.confirmPassword?.length > 0
  ) {
    return true;
  }

  return false;
};
