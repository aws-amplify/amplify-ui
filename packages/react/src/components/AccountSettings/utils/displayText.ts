import { DisplayTextTemplate } from '@aws-amplify/ui';

// ChangePassword
export type ChangePasswordDisplayText = DisplayTextTemplate<{
  confirmPasswordFieldLabel?: string;
  currentPasswordFieldLabel?: string;
  newPasswordFieldLabel?: string;
  updatePasswordButtonText?: string;
}>;

type ChangePasswordDisplayTextDefault = Required<ChangePasswordDisplayText>;

export const defaultChangePasswordDisplayText: ChangePasswordDisplayTextDefault =
  {
    confirmPasswordFieldLabel: 'Confirm Password',
    currentPasswordFieldLabel: 'Current Password',
    newPasswordFieldLabel: 'New Password',
    updatePasswordButtonText: 'Update password',
  };

// DeleteUser
export type DeleteUserDisplayText = DisplayTextTemplate<{
  cancelButtonText?: string;
  confirmDeleteButtonText?: string;
  deleteAccountButtonText?: string;
  warningText?: string;
}>;

type DeleteUserDisplayTextDefault = Required<DeleteUserDisplayText>;

export const defaultDeleteUserDisplayText: DeleteUserDisplayTextDefault = {
  cancelButtonText: 'Cancel',
  confirmDeleteButtonText: 'Delete',
  deleteAccountButtonText: 'Delete Account',
  warningText:
    'Deleting your account is not reversible. You will lose access to your account and all data associated with it.',
};
