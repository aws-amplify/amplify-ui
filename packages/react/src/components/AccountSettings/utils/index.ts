export {
  defaultChangePasswordDisplayText,
  defaultDeleteUserDisplayText,
  ChangePasswordDisplayText,
  DeleteUserDisplayText,
} from './displayText';
