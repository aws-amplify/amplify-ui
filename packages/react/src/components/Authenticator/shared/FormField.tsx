import { FormFieldOptions, getErrors } from '@aws-amplify/ui';

import { PasswordField } from '../../../primitives/PasswordField';
import { PhoneNumberField } from '../../../primitives/PhoneNumberField';
import { TextField } from '../../../primitives/TextField';
import { useAuthenticator } from '../hooks/useAuthenticator';
import { ValidationErrors } from './ValidationErrors';

export interface FormFieldProps {
  name: string;
  formFieldOptions: FormFieldOptions;
}
export function FormField({ name, formFieldOptions }: FormFieldProps) {
  const { validationErrors } = useAuthenticator((context) => [
    context.validationErrors,
  ]);
  const { type } = formFieldOptions;

  const errors = getErrors(validationErrors[name]);
  const hasError = errors?.length > 0;

  if (type === 'tel') {
    return (
      <>
        <PhoneNumberField
          name={name}
          label={formFieldOptions.label}
          placeholder={formFieldOptions.placeholder}
          defaultCountryCode={formFieldOptions.dialCode}
          countryCodeName="country_code"
          dialCodeList={formFieldOptions.dialCodeList}
          autoComplete={formFieldOptions.autocomplete}
          isRequired={formFieldOptions.isRequired}
          labelHidden={formFieldOptions.labelHidden}
          hasError={hasError}
        />
        <ValidationErrors errors={errors} />
      </>
    );
  } else if (type === 'password') {
    return (
      <>
        <PasswordField
          name={name}
          label={formFieldOptions.label}
          autoComplete={formFieldOptions.autocomplete}
          placeholder={formFieldOptions.placeholder}
          isRequired={formFieldOptions.isRequired}
          labelHidden={formFieldOptions.labelHidden}
          hasError={hasError}
        />
        <ValidationErrors errors={errors} />
      </>
    );
  } else {
    return (
      <>
        <TextField
          name={name}
          label={formFieldOptions.label}
          placeholder={formFieldOptions.placeholder}
          autoComplete={formFieldOptions.autocomplete}
          isRequired={formFieldOptions.isRequired}
          labelHidden={formFieldOptions.labelHidden}
          hasError={hasError}
          type={type}
        />
        <ValidationErrors errors={errors} />
      </>
    );
  }
}
