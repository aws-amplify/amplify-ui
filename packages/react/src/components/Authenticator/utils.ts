export const isSignInOrSignUpRoute = (route: string): boolean =>
  route === 'signIn' || route === 'signUp';
