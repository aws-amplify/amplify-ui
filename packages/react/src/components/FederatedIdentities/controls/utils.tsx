import {
  ProviderData,
  ProviderType,
  DefaultFederatedProviderList,
} from './types';
import { capitalize, FederatedProvider } from '@aws-amplify/ui';

function getSupportedProviderData<T extends string = string>(
  providerName: T
): ProviderData<T> {
  return {
    displayName: capitalize<T>(providerName),
    icon: providerName,
    providerName: providerName,
  };
}

function validateProviderTypes(providers: ProviderType[]): void {
  const providerNames = new Set<string>();

  providers.forEach((provider) => {
    const providerName =
      typeof provider === 'string' ? provider : provider.providerName;

    if (providerNames.has(providerName)) {
      throw new Error(`Duplicate provider name found: ${providerName}`);
    }

    providerNames.add(providerName);
  });
}

export function toProviderData<T extends string = string>(
  providers: ProviderType<T>[]
): ProviderData<T>[] {
  validateProviderTypes(providers);
  return providers.map((provider) => {
    if (DefaultFederatedProviderList.includes(provider as FederatedProvider)) {
      return getSupportedProviderData<T>(provider as T);
    } else {
      return provider as ProviderData<T>;
    }
  });
}
