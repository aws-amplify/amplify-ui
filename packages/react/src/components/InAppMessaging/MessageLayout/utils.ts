import tinycolor from 'tinycolor2';

import { MessageComponentStyles } from '../hooks/useMessageProps';

import { MessageLayoutButtonModifier } from './types';

/**
 * Utility for determining the class modifier for an In-App Messaging button
 * based on its background color
 *
 * @param buttonStyles button styles which should contain the background color
 *
 * @returns the modifier - either 'light' or 'dark'
 */
export const getButtonModifier = (
  buttonStyles: MessageComponentStyles['primaryButton' | 'primaryButton']
): MessageLayoutButtonModifier => {
  const { backgroundColor } = buttonStyles ?? {};
  const color = tinycolor(backgroundColor);
  return color.isDark() ? 'dark' : 'light';
};
