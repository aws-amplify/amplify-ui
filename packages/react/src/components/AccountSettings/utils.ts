import { PasswordValidator } from './ChangePassword/types';

export const runCustomValidators = (
  validators: PasswordValidator[],
  password: string
): string[] => {
  const errors: string[] = [];
  validators.forEach((validator) => {
    const error = validator(password);
    if (error) {
      errors.push(error);
    }
  });
  return errors.length > 0 ? errors : null;
};
