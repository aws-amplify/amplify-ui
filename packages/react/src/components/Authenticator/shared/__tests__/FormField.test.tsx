import { render, screen } from '@testing-library/react';
import * as React from 'react';

import { defaultFormFieldOptions } from '@aws-amplify/ui';
import { FormField, FormFieldProps } from '../FormField';
import * as UIReactCore from '@aws-amplify/ui-react-core';
import * as UseStableIdModule from '../../../../primitives/utils/useStableId';

jest.mock('@aws-amplify/ui-react-core');

const useStableIdSpy = jest.spyOn(UseStableIdModule, 'useStableId');

const useAuthenticatorSpy = jest.spyOn(UIReactCore, 'useAuthenticator');

const ERROR_MESSAGE = 'Username cannot be empty.';

const usernameFormFieldProps: FormFieldProps = {
  ...defaultFormFieldOptions.username,
  name: 'username',
  label: 'Username',
};

describe('FormField', () => {
  afterEach(() => {
    useAuthenticatorSpy.mockRestore();
    useStableIdSpy.mockRestore();
  });

  it('renders as expected in the happy path', () => {
    // mocking ids so snapshots are consistent
    useStableIdSpy.mockReturnValue('mock-id-0');
    useAuthenticatorSpy.mockReturnValue({
      validationErrors: {},
    } as UIReactCore.UseAuthenticator);

    const { container } = render(<FormField {...usernameFormFieldProps} />);
    expect(container).toMatchSnapshot();
  });

  it('renders as expected in the unhappy path', () => {
    useStableIdSpy.mockReturnValue('mock-id-0');
    useAuthenticatorSpy.mockReturnValue({
      validationErrors: { username: ERROR_MESSAGE },
    } as unknown as UIReactCore.UseAuthenticator);

    const { container } = render(<FormField {...usernameFormFieldProps} />);
    expect(container).toMatchSnapshot();
  });

  it('renders with autocapitalize disabled by default', async () => {
    useStableIdSpy.mockReturnValue('mock-id-0');
    useAuthenticatorSpy.mockReturnValue({
      validationErrors: {},
    } as UIReactCore.UseAuthenticator);

    const { container } = render(<FormField {...usernameFormFieldProps} />);
    const inputField = await screen.findByLabelText('Username');
    expect(inputField.getAttribute('autocapitalize')).toBe('off');
  });

  it('Invalid field is described by validation error', async () => {
    useAuthenticatorSpy.mockReturnValue({
      validationErrors: { username: ERROR_MESSAGE },
    } as unknown as UIReactCore.UseAuthenticator);

    const { container } = render(<FormField {...usernameFormFieldProps} />);

    const inputField = await screen.findByLabelText('Username');
    expect(inputField).toHaveAttribute('aria-describedBy');

    const describedById = inputField.getAttribute('aria-describedBy');

    // now, follow the id this id just like a screen reader would
    const errors = container.querySelector(`[id="${describedById}"]`);

    expect(errors).toHaveTextContent(ERROR_MESSAGE);
  });
});
