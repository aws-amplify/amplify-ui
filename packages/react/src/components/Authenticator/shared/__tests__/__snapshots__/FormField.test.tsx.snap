// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField renders as expected in the happy path 1`] = `
<div>
  <div
    class="amplify-flex amplify-field amplify-textfield"
  >
    <label
      class="amplify-label"
      for="mock-id-0"
    >
      Username
    </label>
    <div
      class="amplify-flex amplify-field-group amplify-field-group--horizontal"
    >
      <div
        class="amplify-field-group__field-wrapper amplify-field-group__field-wrapper--horizontal"
      >
        <input
          aria-invalid="false"
          autocapitalize="off"
          autocomplete="username"
          class="amplify-input amplify-field-group__control"
          id="mock-id-0"
          name="username"
          placeholder="Enter your Username"
          required=""
          type="text"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`FormField renders as expected in the unhappy path 1`] = `
<div>
  <div
    class="amplify-flex amplify-field amplify-textfield"
  >
    <label
      class="amplify-label"
      for="amplify-id-:r1:"
    >
      Username
    </label>
    <div
      class="amplify-flex amplify-field-group amplify-field-group--horizontal"
    >
      <div
        class="amplify-field-group__field-wrapper amplify-field-group__field-wrapper--horizontal"
      >
        <input
          aria-describedby="amplify-id-:r0:"
          aria-invalid="true"
          autocapitalize="off"
          autocomplete="username"
          class="amplify-input amplify-field-group__control amplify-input--error"
          id="amplify-id-:r1:"
          name="username"
          placeholder="Enter your Username"
          required=""
          type="text"
        />
      </div>
    </div>
  </div>
  <div
    data-amplify-sign-up-errors="true"
    id="amplify-id-:r0:"
  >
    <p
      class="amplify-text amplify-text--error"
      role="alert"
    >
      Username cannot be empty.
    </p>
  </div>
</div>
`;
